minicbor-serde: add support for serializing / deserializing custom CBOR tags

mpanav opened the pull request:
Patch series adds support to deserialize / serialize custom tags in minicbor-serde. 

See #44 for more details

mpanav committed:
introduce tag module

Patch introduces wrapper tag types to enable deserialization of tagged values.

Signed-off-by: Panav Munshi <mpanav@amazon.com>

mpanav committed:
de: add support to deserialize into tag wrappers

Signed-off-by: Panav Munshi <mpanav@amazon.com>

mpanav committed:
ser: add support to serialize tag wrappers

Signed-off-by: Panav Munshi <mpanav@amazon.com>

twittner commented on minicbor-serde/src/tag.rs:
```diff
@@ -0,0 +1,255 @@
+/// custom aliases to ensure that [`crate::de::Deserializer`] is able to drive
+/// the [`minicbor::Decoder`] correctly.
+enum TagContainer<T> {
+    Tag(u64, T),
```
```suggestion
    Tag(Tag, T),
```

twittner commented on minicbor-serde/src/tag.rs:
```diff
@@ -0,0 +1,255 @@
+/// Helper struct for formatting tag mismatch errors
+struct TagMismatchError {
+    found: u64,
+    expected: u64,
```
```suggestion
    found: Tag,
    expected: Tag,
```

twittner reviewed (commented):
Thanks! Overall this looks good to me. Just left some minor comments.

mpanav commented on minicbor-serde/src/ser.rs:
```diff
@@ -340,7 +348,17 @@ where
     fn serialize_field<T: Serialize + ?Sized>(&mut self, x: &T) -> Result<(), Self::Error> {
-        x.serialize(&mut *self.serializer)
+        match self.tagged {
+            true => {
```
Sounds good

mpanav commented on minicbor-serde/src/de.rs:
```diff
@@ -426,3 +437,124 @@ impl<'a, 'de> VariantAccess<'de> for Enum<'a, 'de> {
 }
+
+enum State {
+    Tag(u64),
```
Didn't think of it :) Will change this to use `minicbor::data::Tag`

mpanav commented on minicbor-serde/src/de.rs:
```diff
@@ -426,3 +437,124 @@ impl<'a, 'de> VariantAccess<'de> for Enum<'a, 'de> {
+    {
+        match self.state {
+            State::Tag(tag) => {
+                self.state = State::Value;
```
Yup as we'd error out either way. I'd keep it this way as otherwise the code ends up looking a bit too verbose (assign to var, switch state, and then return the value).

Let me know if you feel otherwise.

mpanav committed:
add interop unit tests for tags

Signed-off-by: Panav Munshi <mpanav@amazon.com>

mpanav commented:
Apologies that I wasn't able to get to this in time - work got busy :) I've relanded the commits after having addressed your feedback @twittner.

Let me know if you want me to change anything. Thanks!

twittner committed:
Merge branch 'develop' into custom-tags

twittner commented:
Thanks @mpanav, I finally got around to merge your PR. `minicbor-serde-v0.7.0` has been published with this change to crates.io.
